tensorflow_addons/metrics/map.py

```diff
@@ -0,0 +1,166 @@
+        super(MAP, self).__init__(name=name)
+        if iou_threshold < 0.0 or iou_threshold > 1.0:
+            raise ValueError(
+                "iou threshold value should be greater equal than zero and less equal than one"
+            )
+
+        self.num_classes = num_classes
+        self.iou_threshold = iou_threshold
+        self.accumulate_true_positives = self.add_weight(
+            'accumulate_true_positives',
+            shape=[self.num_classes],
+            initializer='zeros',
+            dtype=tf.float32)
+        self.accumulate_false_positives = self.add_weight(
+            'accumulate_false_positives',
+            shape=[self.num_classes],
+            initializer='zeros',
+            dtype=tf.float32)
+        self.accumulate_ground_truth = self.add_weight(
+            'accumulate_ground_truth',
+            shape=[self.num_classes],
+            initializer='zeros',
+            dtype=tf.float32)
+
+    def update_state(self, y_true, y_pred):
+        y_true = tf.cast(y_true, tf.float32)
+        y_pred = tf.cast(y_pred, tf.float32)
+        total_tp_count = tf.constant([], shape=[0])
+        total_fp_count = tf.constant([], shape=[0])
+        total_ground_truth_count = tf.constant([], shape=[0])
+        for class_index in range(self.num_classes):
+            y_true_class = tf.gather(y_true,
+                                     tf.where(y_true[:, 0] == class_index))
+            y_pred_class = tf.gather(y_pred,
+                                     tf.where(y_pred[:, 0] == class_index))
+            y_true_boxes = y_true_class[:, 1:5]
+            y_pred_boxes = y_pred_class[:, 2:]
+
+            def calcluate_iou(y_true_box):
+                ious = self.iou(tf.expand_dims(y_true_box, 0), y_pred_boxes)
+                ious = tf.gather(ious, tf.where(ious > self.iou_threshold))
+                if tf.equal(tf.shape(ious), 0):
+                    return -1
+                return tf.arg_max(ious)
+
+            iou_result = tf.map_fn(calcluate_iou, y_true_boxes)
+            tp = tf.cast(iou_result >= 0, tf.float32)
+            tp_count = tf.math.count_nonzero(tp)
+            fp_count = tf.shape(tp)[0] - tp_count
+            ground_truth_count = tf.shape(y_true)[0]
+            total_tp_count = tf.concat([total_tp_count, [tp_count]], 0)
+            total_fp_count = tf.concat([total_fp_count, [fp_count]], 0)
+            total_ground_truth_count = tf.concat(
+                [total_ground_truth_count, [ground_truth_count]], 0)
+
+        self.accumulate_true_positives.assign_add(total_tp_count)
+        self.accumulate_false_positives.assign_add(total_fp_count)
+        self.accumulate_ground_truth.assign_add(total_ground_truth_count)
+
+    def iou(self, b1, b2):
+        """
+        Args
+            b1: bbox.
+            b1: the other bbox.
+        Returns:
+            iou float `Tensor`.
+        """
+        b1_ymin = tf.minimum(b1[:, 0], b1[:, 2])
```

Comment:
It looks like this line fails with when I try to run the unittests
```
InvalidArgumentError: slice index 0 of dimension 1 out of bounds. [Op:StridedSlice] name: map/while/strided_slice/
```

Reply:
I will fix it later

Reply:
Thanks for taking a look

Reply:
Any updates on this?

Reply:
Here is my [implementation](https://github.com/fsx950223/mobilenetv2-yolov3/blob/d8ecb19b8ffd6f9f2bde6d313de609eebd44614a/yolo3/map.py#L120) of map.You can also revert to first commit to getting the runnable version.